packages/eslint-plugin/src/rules/no-extraneous-return-types.ts

```diff
@@ -0,0 +1,311 @@
+import type { TSESTree } from '@typescript-eslint/utils';
+import type * as ts from 'typescript';
+
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+
+import { createRule, forEachReturnStatement, getParserServices } from '../util';
+
+export default createRule({
+  name: 'no-extraneous-return-types',
+  meta: {
+    type: 'suggestion',
+    docs: {
+      description: 'Disallow extraneous union types in return type annotations',
+      requiresTypeChecking: true,
+    },
+    fixable: 'code',
+    messages: {
+      unusedReturnTypes:
+        "Unused or redundant union type '{{type}}' in return type.",
+    },
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const services = getParserServices(context);
+    const checker = services.program.getTypeChecker();
+
+    function getRelevantReturnTypes(
```

Comment:
[Style] Nit: this function both _gets_ and _checks_. It took me a re-read to get that it does both. WDYT about expanding its name to mention both?

```suggestion
    function getRelevantReturnTypesOrReport(
```